src/QMCWaveFunctions/Fermion/MatrixUpdateOMPTarget.h

```diff
@@ -197,6 +199,69 @@ class MatrixUpdateOMPTarget
       grad_now[iw] = {grads_value_v[iw][0], grads_value_v[iw][1], grads_value_v[iw][2]};
   }
 
+  template<typename GT>
+  static void mw_evalGradWithSpin(const RefVectorWithLeader<This_t>& engines,
+                                  const std::vector<const Value*>& dpsiM_row_list,
+                                  const std::vector<const Value*>& dspin_row_list,
+                                  int rowchanged,
+                                  std::vector<GT>& grad_now,
+                                  std::vector<Value>& spingrad_now)
+  {
+    auto& engine_leader     = engines.getLeader();
+    auto& buffer_H2D        = engine_leader.mw_mem_->buffer_H2D;
+    auto& grads_value_v     = engine_leader.mw_mem_->grads_value_v;
+    auto& spingrads_value_v = engine_leader.mw_mem_->spingrads_value_v;
+
+    const int norb                   = engine_leader.get_psiMinv().rows();
+    const int nw                     = engines.size();
+    constexpr size_t num_ptrs_packed = 3; // it must match packing and unpacking
+    buffer_H2D.resize(sizeof(Value*) * num_ptrs_packed * nw);
+    Matrix<const Value*> ptr_buffer(reinterpret_cast<const Value**>(buffer_H2D.data()), num_ptrs_packed, nw);
+    for (int iw = 0; iw < nw; iw++)
+    {
+      ptr_buffer[0][iw] = engines[iw].get_psiMinv().device_data() + rowchanged * engine_leader.get_psiMinv().cols();
+      ptr_buffer[1][iw] = dpsiM_row_list[iw];
+      ptr_buffer[2][iw] = dspin_row_list[iw];
+    }
+
+    constexpr unsigned DIM = GT::Size;
+    grads_value_v.resize(nw, DIM);
+    spingrads_value_v.resize(nw);
+    auto* __restrict__ grads_value_v_ptr     = grads_value_v.data();
```

Comment:
never like to see the cargo cult __restrict__ all over but I notice it elsewhere in this file now, its not part of C++.  Except in legacy files it is used `#defined` so it can be easily dropped. In principle an ideal optimizer should not need it. Do we have any evidence that omp target does something with this?  That's the only justification I can see for this being here.

Reply:
I was basically following what was done elsewhere in the file. I'm not sure exactly what omp target does with it. @ye-luo?